library/core/resource-loader/src/main/java/org/quiltmc/qsl/resource/loader/impl/ResourceLoaderImpl.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.resource.loader.impl;
+
+import java.net.URL;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import net.fabricmc.loader.api.FabricLoader;
+import net.fabricmc.loader.api.ModContainer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import net.minecraft.resource.DefaultResourcePack;
+import net.minecraft.resource.ResourceReloader;
+import net.minecraft.resource.ResourceType;
+import net.minecraft.util.Identifier;
+
+import org.quiltmc.qsl.resource.loader.api.ResourceLoader;
+import org.quiltmc.qsl.resource.loader.api.ResourcePackActivationType;
+import org.quiltmc.qsl.resource.loader.api.reloader.IdentifiableResourceReloader;
+
+/**
+ * Represents the implementation of the resource loader.
+ */
+public final class ResourceLoaderImpl implements ResourceLoader {
+	private static final Map<ResourceType, ResourceLoaderImpl> IMPL_MAP = new HashMap<>();
+	private static final Logger LOGGER = LogManager.getLogger();
+
+	private final Set<Identifier> addedListenerIds = new HashSet<>();
+	private final Set<IdentifiableResourceReloader> addedReloaders = new LinkedHashSet<>();
+
+	public static ResourceLoaderImpl get(ResourceType type) {
+		return IMPL_MAP.computeIfAbsent(type, t -> new ResourceLoaderImpl());
+	}
+
+	public static void sort(ResourceType type, List<ResourceReloader> reloaders) {
+		get(type).sort(reloaders);
+	}
+
+	@Override
+	public void registerReloader(IdentifiableResourceReloader resourceReloader) {
+		if (!this.addedListenerIds.add(resourceReloader.getQuiltId())) {
+			throw new RuntimeException("Tried to register resource reloader " + resourceReloader.getQuiltId() + " twice!");
```

Comment:
There may be a better exception type than runtime exception for these? Runtime exception is quite broad

Reply:
Replaced with IllegalStateExceptions.